### ISSUE TITLE
Rule Change: add warning on obsolete use of `eslint-env` configuration comments in 9.x

### ISSUE THREAD

**Issue by MikeMcC399** [eslint/eslint#20340]
### What rule do you want to change?

https://github.com/eslint/eslint/blob/3383e7ec9028166cafc8ea7986c2f7498d0049f0/lib/languages/js/source-code/source-code.js#L1051-L1059

### What change do you want to make?

Implement suggestions

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint-env node */
```

### What does the rule currently do for this code?

In ESLint 10 alpha it produces an error message:

>   1:1  error  /* eslint-env */ comments are no longer supported

In ESLint 9.36.1 no error or warning is shown

### What will the rule do after it's changed?

It will cause a warning if an [`eslint-env` Configuration Comments](https://eslint.org/docs/latest/use/configure/migration-guide#eslint-env-configuration-comments) is used in ESLint 9.x in a flat configuration environment.

Edit: discussions led to the modified suggestion:

Emit a warning using `process.emitWarning()` to avoid a potential breaking change of using a lint warning, if [--max-warnings](https://eslint.org/docs/latest/use/command-line-interface#--max-warnings) were used.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by JoshuaKGoldberg** [comment]
👍 I'm generally in favor of this. ESLint v9 obviously can't _error_ on them, as that'd be a breaking change. But providing some kind of warning in flat config mode to let folks know they do nothing in flat config would help surface that information.

That being said, I wonder if some users would receive this as an effective "breaking" change? They might have their CI set up to fail on warnings, or just generally see new classifications/swathes of console output as too significant. A couple of ideas:

* Adding an environment variable and/or flag to opt out of these warnings
* Automatically removing them when in `--fix` mode

**Comment by MikeMcC399** [comment]
@JoshuaKGoldberg 

> * Automatically removing them when in `--fix` mode

Currently in `eslint@10.0.0-alpha.0`, `npx eslint --fix` does not remove an `eslint-env` comment. That would be nice to have, however I think that enhancement consideration should probably go into its own separate issue. It could then be handled consistently across ESLint 9.x and 10.x.

**Comment by JoshuaKGoldberg** [comment]
Not necessarily. I think there's a possibility that some users would consider this to be a painful change, and the only way that they wouldn't be absolutely enraged & inconvenienced by it would be if ESLint provided an automatic way to fix the complaints.

However, if the team doesn't believe that to be the case, I agree - adding a fixer would be better as a separate feature request.

**Comment by mdjermanovic** [comment]
Removing `eslint-env` comments with `--fix` makes sense when they're just leftovers after the migration, but we can't know if that's the case. Generally, removing `eslint-env` comments is part of the eslintrc -> flat config migration process that requires additional work of adding globals, either inline or in the config file:

https://eslint.org/docs/latest/use/configure/migration-guide#eslint-env-configuration-comments

So I don't think we should automatically remove these comments, but just report them since we don't know if an additional work needs to be done. `--fix` should generally leave the code in a valid, "fixed" state in regard to the problems it fixed, which isn't necessarily the case here.

**Comment by mdjermanovic** [comment]
As for adding warnings in ESLint v9, the v9.x line is now in the Maintenance status, which means it receives only critical bug fixes and compatibility fixes to ensure interoperability:

https://eslint.org/version-support/

This isn't a critical bug fix, and I don't think it qualifies as a compatibility fix to ensure interoperability. @eslint/eslint-tsc thoughts?

**Comment by MikeMcC399** [comment]
@mdjermanovic 

> This isn't a critical bug fix, and I don't think it qualifies as a compatibility fix to ensure interoperability.

I understand your point. I did raise this first in Discord to ask if I should submit an issue, and @nzakas recommended posting it for discussion.

The impact of not implementing the suggestion is simply that users may get an error (fatal) when upgrading to ESLint 10, whereas if they were warned (non fatal) in ESLint 9, they could clean up the sources ahead of time, making the upgrade transition perhaps a little smoother.

I won't object if you and the team decide to close without implementing!

**Comment by nzakas** [comment]
Yeah, I suggested opening an issue to see what the team thought. This is one of those cases where I wonder if warning in v9.x would provide a softer landing for v10.x. There are undoubtedly folks who have `/* eslint-env */` still in their files and will be surprised by an error when they upgrade. Arguably, we should have provided this error in flat config mode in v9.x already, so I view it as a compatibility issue. We can't make it an error in v9.x because that would be a breaking change but emitting as a warning seems reasonable.

Happy to defer to consensus on this.

**Comment by mdjermanovic** [comment]
Does proposed warning means a lint message with warning severity (`severity: 1`) or `process.emitWarning()`?

**Comment by MikeMcC399** [comment]
> Does proposed warning means a lint message with warning severity (`severity: 1`) or `process.emitWarning()`?

The intention of my proposal would be to have something similar for ESLint 9.x, in other words, a lint message warning like:

> 1:1 warning /* eslint-env */ comments are no longer supported in flat configurations

This should not be triggered if ESLint 9.x is running with a legacy `eslintrc` configuration.

**Comment by nzakas** [comment]
This is what we're doing in v10:
https://github.com/eslint/eslint/blob/0f5a94a84beee19f376025c74f703f275d52c94b/lib/languages/js/source-code/source-code.js#L1052-L1057

IIRC, problems are always reported as errors. 🤔 So we might need to do it as `process.emitWarning()` in v9.x.

**Comment by mdjermanovic** [comment]
I'm in favor of `process.emitWarning()`, regardless of technical problems we'd have with implementing lint message warnings, because new lint message warnings could be seen as a breaking change, e.g., when [`--max-warnings`](https://eslint.org/docs/latest/use/command-line-interface#--max-warnings) is used.

**Comment by nzakas** [comment]
That makes sense. It sounds like we've agreed to emit a warning using `process.emitWarning()` in v9.x, so marking as accepted.

**Comment by sethamus** [comment]
I will work on this.

**Comment by nzakas** [comment]
@sethamus thanks!

**Comment by mdjermanovic** [comment]
Just to note that the PR should target `v9.x-dev` branch.

**Comment by MikeMcC399** [comment]
Implemented in release [v9.39.2](https://github.com/eslint/eslint/releases/tag/v9.39.2)
